File: app/src/main/java/ru/santaev/clipboardtranslator/TranslatorApp.java
package ru.santaev.clipboardtranslator;


import android.app.Application;
import android.arch.persistence.room.Room;
import android.content.Context;

import ru.santaev.clipboardtranslator.api.ApiService;
import ru.santaev.clipboardtranslator.db.AppDatabase;
import ru.santaev.clipboardtranslator.model.DataModel;
import ru.santaev.clipboardtranslator.model.HistoryDataModel;
import ru.santaev.clipboardtranslator.model.IDataModel;
import ru.santaev.clipboardtranslator.model.IHistoryDataModel;

public class TranslatorApp extends Application{

    private static TranslatorApp instance;

    public static TranslatorApp getInstance(){
        return instance;
    }

    public static Context getAppContext(){
        return getInstance().getApplicationContext();
    }

    private IDataModel dataModel;
    private IHistoryDataModel historyDataModel;
    private AppDatabase appDatabase;

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;

        appDatabase = buildDatabase();
        dataModel = buildDataModel();
        historyDataModel = buildHistoryDataModel();
    }

    public IDataModel getDataModel() {
        return dataModel;
    }

    public AppDatabase getDatabase() {
        return appDatabase;
    }

    public IHistoryDataModel getHistoryDataModel() {
        return historyDataModel;
    }

    protected IDataModel buildDataModel(){
        return new DataModel(new ApiService());
    }

    protected IHistoryDataModel buildHistoryDataModel() {
        return new HistoryDataModel();
    }

    protected AppDatabase buildDatabase() {
        return Room.databaseBuilder(TranslatorApp.getAppContext(), AppDatabase.class, AppDatabase.DB_NAME)
                .build();
    }
}

File: app/src/main/java/ru/santaev/clipboardtranslator/TranslatorApp.kt
package ru.santaev.clipboardtranslator


import android.app.Application
import android.content.Context
import com.example.santaev.domain.factory.IGatewayFactory
import ru.santaev.clipboardtranslator.di.AppComponent
import ru.santaev.clipboardtranslator.di.DaggerAppComponent
import ru.santaev.clipboardtranslator.di.module.AppModule


open class TranslatorApp : Application() {

    lateinit var appComponent: AppComponent
        private set

    override fun onCreate() {
        super.onCreate()
        instance = this

        appComponent = buildComponent()

        IGatewayFactory.setGatewayFactory(GatewayFactory(
                appComponent.getDatabase().getLanguageDao(),
                appComponent.getDatabase().getTranslationDao(),
                appComponent.getApiService()
        ))
    }

    protected open fun buildComponent(): AppComponent {
        return DaggerAppComponent.builder()
                .appModule(AppModule(this, BuildConfig.MOCK))
                .build()
    }

    companion object {

        @JvmStatic
        lateinit var instance: TranslatorApp
            private set

        val appContext: Context
            get() = instance.applicationContext
    }
}
